l1-contracts/contracts/chain-registrar/ChainRegistrar.sol

```diff
@@ -0,0 +1,161 @@
+    /// Address that will be used for deploying l2 contracts
+    address public l2Deployer;
+    /// ZKsync Bridgehub
+    IBridgehub public bridgehub;
+
+    /// Chains that has been successfully deployed
+    mapping(bytes32 => bool) public deployedChains;
+    /// Proposal for chain registration
+    mapping(bytes32 => ChainConfig) public proposedChains;
+
+    error ProposalNotFound();
+    error BaseTokenTransferFailed();
+    error ChainIsAlreadyDeployed();
+    error ChainIsNotYetDeployed();
+    error BridgeIsNotRegistered();
+
+    /// @notice new chain is deployed
+    event NewChainDeployed(uint256 indexed chainId, address author, address diamondProxy, address chainAdmin);
+
+    /// @notice new chain is proposed to register
+    event NewChainRegistrationProposal(uint256 indexed chainId, address author, bytes32 key);
+
+    /// @notice Shared bridge is registered on l2
+    event SharedBridgeRegistered(uint256 indexed chainId, address l2Address);
+
+    /// @notice L2 Deployer has changed
+    event L2DeployerChanged(address newDeployer);
+
+    struct BaseToken {
+        address tokenAddress;
+        address tokenMultiplierSetter;
+        uint128 gasPriceMultiplierNominator;
+        uint128 gasPriceMultiplierDenominator;
+    }
+
+    struct ChainConfig {
+        /// @param Chain id of the new chain should be unique for this bridgehub
+        uint256 chainId;
+        /// @param Operator for making commit txs.
+        address commitOperator;
+        /// @param Operator for making Prove and Execute transactions
+        address operator;
+        /// @param Governor of the chain. Ownership of the chain will be transferred to this operator
+        address governor;
+        /// @param baseToken of the chain
+        BaseToken baseToken;
+        /// @param pubdataPricingMode How the users will charged for pubdata.
+        PubdataPricingMode pubdataPricingMode;
+    }
+
+    constructor(address _bridgehub, address _l2Deployer, address _owner) reentrancyGuardInitializer {
+        bridgehub = IBridgehub(_bridgehub);
+        l2Deployer = _l2Deployer;
+        _transferOwnership(_owner);
+    }
+
+    function proposeChainRegistration(
+        uint256 chainId,
+        PubdataPricingMode pubdataPricingMode,
+        address commitOperator,
+        address operator,
+        address governor,
+        address tokenAddress,
+        address tokenMultiplierSetter,
+        uint128 gasPriceMultiplierNominator,
+        uint128 gasPriceMultiplierDenominator
+    ) public {
+        ChainConfig memory config = ChainConfig({
+            chainId: chainId,
+            pubdataPricingMode: pubdataPricingMode,
+            commitOperator: commitOperator,
+            operator: operator,
+            governor: governor,
+            baseToken: BaseToken({
+                tokenAddress: tokenAddress,
+                tokenMultiplierSetter: tokenMultiplierSetter,
+                gasPriceMultiplierNominator: gasPriceMultiplierNominator,
```

Comment:
maybe adding more validations on `gasPriceMultiplierNominator`, I'd say at min check both values are 1, 1 if base token is ETH